velox/functions/lib/Re2Functions.cpp

```diff
@@ -383,13 +383,220 @@ class Re2SearchAndExtract final : public VectorFunction {
   const bool emptyNoMatch_;
 };
 
+namespace {
+
+// Sub-pattern's stats in a top-level pattern.
+struct SubPatternStats {
+  SubPatternKind kind;
+  // Count of this patterns in total.
+  size_t count = 0;
+  // First index of this pattern kind.
+  int32_t firstIndex = -1;
+  // Last index of this pattern kind.
+  int32_t lastIndex = -1;
+
+  void update(size_t index) {
+    count++;
+    if (firstIndex == -1) {
+      firstIndex = index;
+    }
+
+    lastIndex = index;
+  }
+};
+
+// Construct SubPatterns from subPatternKinds, subPatternRanges.
+// Caller need to make sure the specified range only contains
+// fixed(kLiteralString, kSingleWildcard) patterns.
+size_t buildFixedSubPatterns(
+    const std::vector<SubPatternKind>& subPatternKinds,
+    const std::vector<std::pair<size_t, size_t>>& subPatternRanges,
+    size_t start,
+    size_t end,
+    std::vector<SubPatternMetadata>& subPatterns) {
+  size_t fixedLength = 0;
+  for (auto i = start; i < end; i++) {
+    if (subPatternKinds[i] == SubPatternKind::kLiteralString) {
```

Comment:
the next functions uses a single 'if'

`if (subPatternKinds[i] != SubPatternKind::kAnyCharsWildcard) {`

any particular reason not to do the same here

it is hard to read the code when it is not consistent

Reply:
Would it make sense to combine 2 "if" branches?

```
const auto kind = subPatternKinds[i];
if (kind == SubPatternKind::kLiteralString || kind == SubPatternKind::kSingleCharWildcard) {
   subPatterns.push_back(
          {kind,
           indexInFixedPattern,
           subPatternRanges[i].second});
}
```